README.md
# Introductions

This repo contains a Haskell implementation of the ColonyRuntime API, making it possible to implement Colonies applications/workers in Haskell.

## Example code
The code below assigns a Colonies process and calculates the last number in a Fibonacci series. 
```haskell
module Main where

import CryptoLib
import ColoniesLib
import Control.Concurrent (threadDelay)
import Control.Monad (forever)

colonyId = "4787a5071856a4acf702b2ffcea422e3237a679c681314113d86139461290cf4"
runtimePrvKey = "ddf7f7791208083b6a9ed975a72684f6406a269cfa36f1b1c32045c0a71fff05"
host = "http://localhost:50080"

fib 0 = 0
fib 1 = 1
fib n = fib (n-1) + fib (n-2)

worker :: IO ()
worker = do
    -- Connect to the Colonies server and try to assign a process to execute from the job queue
    maybeProcess <- assign colonyId host runtimePrvKey
    if maybeProcess == Nothing then
        threadDelay 1000000 -- Job queue is empty, wait 1 second and try again ...
    else do
        -- Parse process parameters
        let process = maybe createEmptyProcess id maybeProcess
        cmd <- getCmd process
        args <- getArgs process
        if cmd == "fibonacci" then do
            print "Got a Fibonacci task"
            let n = read $ head args :: Integer
            let f = fib n 
            print $ "-> fib " ++ show n ++ "=" ++ show f
            -- Create a an attribute 
            let attr = createAttribute process "output" $ show f
            -- Connect to the Colnies server and add the add attribute to the process object
            -- Note: addAddtribute will return Nothing in case of error, else a Just Attribute
            maybeAddedAttr <- addAttribute attr host runtimePrvKey
            close process host runtimePrvKey 
            print "Done calculating Fibonacci"
        else
            print "Invalid cmd"

main :: IO ()
main = forever worker 
```

## To test it ... 
### Start a Colonies dev server 
```console
colonies dev
```

Start a Haskell worker (the example code). Note, you need to first install Stack and clone the repo.
```console
stack run
```

### To fully use the colonies CLI, the following environmental variables needs to be added
```console
export LANG=en_US.UTF-8
export LANGUAGE=en_US.UTF-8
export LC_ALL=en_US.UTF-8
export LC_CTYPE=UTF-8
export TZ=Europe/Stockholm
export COLONIES_TLS="false"
export COLONIES_SERVERHOST="localhost "
export COLONIES_SERVERPORT="50080"
export COLONIES_MONITORPORT="21120"
export COLONIES_MONITORINTERVALL="1"
export COLONIES_SERVERID="039231c7644e04b6895471dd5335cf332681c54e27f81fac54f9067b3f2c0103"
export COLONIES_SERVERPRVKEY="fcc79953d8a751bf41db661592dc34d30004b1a651ffa0725b03ac227641499d"
export COLONIES_DBHOST="localhost"
export COLONIES_DBUSER="postgres"
export COLONIES_DBPORT="50070"
expoer COLONIES_DBPASSWORD="rFcLGNkgsNtksg6Pgtn9CumL4xXBQ7"
export COLONIES_COLONYID="4787a5071856a4acf702b2ffcea422e3237a679c681314113d86139461290cf4"
export COLONIES_COLONYPRVKEY="ba949fa134981372d6da62b6a56f336ab4d843b22c02a4257dcf7d0d73097514"
export COLONIES_RUNTIMEID="3fc05cf3df4b494e95d6a3d297a34f19938f7daa7422ab0d4f794454133341ac"
export COLONIES_RUNTIMEPRVKEY="ddf7f7791208083b6a9ed975a72684f6406a269cfa36f1b1c32045c0a71fff05"
export COLONIES_RUNTIMETYPE="cli"
```

### Submit a process to calculate Fib(10)
```console
colonies process run --cmd fibonacci --args 10 --runtimetype cli
```
#### Output
```console
INFO[0000] Starting a Colonies client                    Insecure=true ServerHost=localhost ServerPort=50080
INFO[0000] Process submitted                             ProcessID=da41231a1f30039edb5677a8084e5f1c605bc54e3e8fdec565170c4e4e8d23be
```
### Lookup the result 
```console
colonies process get --processid da41231a1f30039edb5677a8084e5f1c605bc54e3e8fdec565170c4e4e8d23be
```

#### Output
```console
Process:
+-------------------+------------------------------------------------------------------+
| ID                | da41231a1f30039edb5677a8084e5f1c605bc54e3e8fdec565170c4e4e8d23be |
| IsAssigned        | True                                                             |
| AssignedRuntimeID | 3fc05cf3df4b494e95d6a3d297a34f19938f7daa7422ab0d4f794454133341ac |
| State             | Successful                                                       |
| Priority          | 0                                                                |
| SubmissionTime    | 2022-07-11 23:46:19                                              |
| StartTime         | 2022-07-11 23:46:19                                              |
| EndTime           | 2022-07-11 23:46:19                                              |
| Deadline          | 0001-01-01 01:12:12                                              |
| WaitingTime       | 473.1ms                                                          |
| ProcessingTime    | 9.766ms                                                          |
| Retries           | 0                                                                |
+-------------------+------------------------------------------------------------------+

ProcessSpec:
+-------------+-----------+
| Image       | None      |
| Cmd         | fibonacci |
| Args        | 10        |
| Volumes     | None      |
| Ports       | None      |
| MaxExecTime | -1        |
| MaxRetries  | -1        |
+-------------+-----------+

Conditions:
+-------------+------------------------------------------------------------------+
| ColonyID    | 4787a5071856a4acf702b2ffcea422e3237a679c681314113d86139461290cf4 |
| RuntimeIDs  | None                                                             |
| RuntimeType | cli                                                              |
| Memory      | 0                                                                |
| CPU Cores   | 0                                                                |
| GPUs        | 0                                                                |
+-------------+------------------------------------------------------------------+

Attributes:
+------------------------------------------------------------------+--------+-------+------+
|                                ID                                |  KEY   | VALUE | TYPE |
+------------------------------------------------------------------+--------+-------+------+
| a584cd5c51067bbdc492dba8361e4a362aacc39b0f9726d483741fd962d5b418 | output | 55    | Out  |
+------------------------------------------------------------------+--------+-------+------+
```

### Submit a process to calculate Fib(10) and wait for the result
```console
colonies process run --cmd fibonacci --args 10 --runtimetype cli --wait
```

#### Output
```console
55⏎
```


ChangeLog.md
# Changelog for haskellrt

## Unreleased changes


test/Spec.hs
module Main where

import Test.Tasty
import Test.Tasty.HUnit
import CryptoLib
import CryptoLibTest
import ColoniesLibTest

main :: IO ()
main = defaultMain tests 

tests :: TestTree
tests = testGroup "Tests" [cryptoLibTests, coloniesLibTest]


test/CryptoLibTest.hs
module CryptoLibTest where

import Test.Tasty
import Test.Tasty.HUnit
import CryptoLib

cryptoLibTests = testGroup "CryptoLib" [testPrvkey,
                                        testIdentity,
                                        testSignature,
                                        testHash,
                                        testRecoverid]

testPrvkey = testCase "prvkey" $ do
    key <- generateKey 
    let l = length key 
    assertBool "expecting length prvkey == 64" $ l == 64

testIdentity = testCase "identity" $ do
    let prvkey = "d6eb959e9aec2e6fdc44b5862b269e987b8a4d6f2baca542d8acaa97ee5e74f6" 
    identity <- identity prvkey
    let expectedIdentity = "5d6568f883451ae2e407d1a0a7992e414f2a67b69d0e6e9176d353b98f06f696"
    assertEqual ("expecting identity == " ++ expectedIdentity) identity expectedIdentity 

testSignature = testCase "sign" $ do
    let prvkey = "d6eb959e9aec2e6fdc44b5862b269e987b8a4d6f2baca542d8acaa97ee5e74f6" 
    signature <- sign "hello" prvkey
    let expectedSignature = "e713a1bb015fecabb5a084b0fe6d6e7271fca6f79525a634183cfdb175fe69241f4da161779d8e6b761200e1cf93766010a19072fa778f9643363e2cfadd640900"
    assertEqual ("expecting signature == " ++ expectedSignature) signature expectedSignature

testHash = testCase "hash" $ do
    digest <- hash "hello"
    let expectedDigest = "3338be694f50c5f338814986cdf0686453a888b84f424d792af4b9202398f392"
    assertEqual ("expecting digest == " ++ expectedDigest) digest expectedDigest

testRecoverid = testCase "recoverid" $ do
    let prvkey = "d6eb959e9aec2e6fdc44b5862b269e987b8a4d6f2baca542d8acaa97ee5e74f6" 
    let signature = "e713a1bb015fecabb5a084b0fe6d6e7271fca6f79525a634183cfdb175fe69241f4da161779d8e6b761200e1cf93766010a19072fa778f9643363e2cfadd640900"
    identity <- recoverid "hello" signature 
    let expectedIdentity = "5d6568f883451ae2e407d1a0a7992e414f2a67b69d0e6e9176d353b98f06f696"
    assertEqual ("expecting identity == " ++ expectedIdentity) identity expectedIdentity


test/ColoniesLibTest.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DuplicateRecordFields #-}

module ColoniesLibTest where

import Test.Tasty
import Test.Tasty.HUnit
import CryptoLib
import ColoniesLib 
import Data.Aeson
import Data.Text as T
import Data.Text.IO as T

serverPrvKey = "fcc79953d8a751bf41db661592dc34d30004b1a651ffa0725b03ac227641499d"
host = "http://localhost:50080"

coloniesLibTest = testGroup "ColoniesLib" [testAddColony, 
                                           testGetColony,
                                           testGetColonies,
                                           testSubmit,
                                           testAssign]

testAddColony = testCase "addColony" $ do
    prvkey <- generateKey
    colonyId <- identity prvkey
    let colony = createColony colonyId "test_colony"
    maybeAddedColony <- addColony colony host serverPrvKey
    let addedColony = maybe Colony { colonyid = "", name="" } id maybeAddedColony
    assertEqual "expecting added colony == colony" colony addedColony

testGetColony = testCase "getColony" $ do
    colonyPrvKey <- generateKey
    colonyId <- identity colonyPrvKey 
    let colony = createColony colonyId "test_colony"
    addColony colony host serverPrvKey
  
    runtimePrvKey <- generateKey
    runtimeId <- identity runtimePrvKey 
    let runtime = createRuntime "test_runtimetype" runtimeId colonyId
    addedRuntime <- addRuntime runtime host colonyPrvKey 
    err <- approveRuntime runtimeId host colonyPrvKey
    assertBool "expecting no error approving runtime" (err==Nothing)
    
    maybeAddedColony <- getColony colonyId host runtimePrvKey 
    let addedColony = maybe Colony { colonyid = "", name="" } id maybeAddedColony
    assertEqual "expecting added colony == colony" colony addedColony

testGetColonies = testCase "getColonies" $ do
    prvkey <- generateKey
    colonyId <- identity prvkey
    let colony = createColony colonyId "test_colony"
    maybeAddedColony <- addColony colony host serverPrvKey
    
    prvkey <- generateKey
    colonyId <- identity prvkey
    let colony = createColony colonyId "test_colony"
    maybeAddedColony <- addColony colony host serverPrvKey
   
    maybeColonies <- getColonies host serverPrvKey
    let colonies = maybe [] id maybeColonies
    let l = Prelude.length colonies 
    assertBool "expecting length colonies>2" (l > 2)

testSubmit = testCase "submit" $ do
    colonyPrvKey <- generateKey
    colonyId <- identity colonyPrvKey 
    let colony = createColony colonyId "test_colony"
    addColony colony host serverPrvKey
  
    runtimePrvKey <- generateKey
    runtimeId <- identity runtimePrvKey 
    let runtime = createRuntime "test_runtimetype" runtimeId colonyId
    addedRuntime <- addRuntime runtime host colonyPrvKey 
    err <- approveRuntime runtimeId host colonyPrvKey
    assertBool "expecting no error approving runtime" (err==Nothing)

    let cond = createConditions colonyId "test_runtimetype" []
    let spec = createProcessSpec "test_process_name" "test_cmd" ["test_arg1", "test_arg2"] 100 3 cond
    let specWithEnv = addEnv spec "test_key" "test_value"  
    maybeProcess <- submit specWithEnv host runtimePrvKey 
    assertBool "expecting process" (maybeProcess/=Nothing)

testAssign = testCase "assign" $ do
    colonyPrvKey <- generateKey
    colonyId <- identity colonyPrvKey 
    let colony = createColony colonyId "test_colony"
    addColony colony host serverPrvKey
  
    runtimePrvKey <- generateKey
    runtimeId <- identity runtimePrvKey 
    let runtime = createRuntime "test_runtimetype" runtimeId colonyId
    addedRuntime <- addRuntime runtime host colonyPrvKey 
    err <- approveRuntime runtimeId host colonyPrvKey
    assertBool "expecting no error approving runtime" (err==Nothing)

    let cond = createConditions colonyId "test_runtimetype" []
    let spec = createProcessSpec "test_process_name" "test_cmd" ["test_arg1", "test_arg2"] 100 3 cond
    let specWithEnv = addEnv spec "test_key" "test_value"  
    maybeSubmittedProcess <- submit specWithEnv host runtimePrvKey 
    let submittedProcess = maybe createEmptyProcess id maybeSubmittedProcess
    assertBool "expecting process" (maybeSubmittedProcess/=Nothing)
    
    maybeAssignedProcess <- assign colonyId host runtimePrvKey
    let assignedProcess = maybe createEmptyProcess id maybeAssignedProcess
    cmd <- getCmd assignedProcess
    args <- getArgs assignedProcess
    assertEqual "expecting correct cmd" cmd "test_cmd"
    assertEqual "expecting correct args" args ["test_arg1", "test_arg2"]


src/CryptoLib.hs
module CryptoLib
    ( generateKey, identity, sign, hash, recoverid
    ) where

{-# LANGUAGE ForeignFunctionInterface #-}
import Foreign
import Foreign.C.Types
import Foreign.C.String

foreign import ccall "prvkey" c_prvkey :: IO CString
foreign import ccall "id" c_identity :: CString -> IO CString
foreign import ccall "sign" c_sign :: CString -> CString -> IO CString
foreign import ccall "hash" c_hash :: CString -> IO CString
foreign import ccall "recoverid" c_recoverid:: CString -> CString -> IO CString

generateKey :: IO String
generateKey = do k <- c_prvkey
                 str <- peekCString k
                 _ <- free k
                 return str 

identity :: String -> IO String
identity prvkey = do prvkeyc <- newCString prvkey 
                     identityc <- c_identity prvkeyc 
                     identity <- peekCString identityc 
                     _ <- free identityc 
                     _ <- free prvkeyc 
                     return identity 

sign :: String -> String -> IO String
sign msg prvkey = do msgc <- newCString msg
                     prvkeyc <- newCString prvkey
                     sigc <- c_sign msgc prvkeyc
                     sig <- peekCString sigc
                     _ <- free sigc
                     _ <- free msgc
                     _ <- free prvkeyc
                     return sig

hash :: String -> IO String
hash msg = do msgc <- newCString msg
              hashc <- c_hash msgc
              hash <- peekCString hashc 
              _ <- free hashc
              _ <- free msgc
              return hash 

recoverid :: String -> String -> IO String
recoverid msg sig = do msgc <- newCString msg
                       sigc <- newCString sig
                       identityc <- c_recoverid msgc sigc
                       identity <- peekCString identityc 
                       _ <- free identityc 
                       _ <- free msgc
                       _ <- free sigc
                       return identity 


src/ColoniesLib.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DataKinds #-}                                                                                                                                                         
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE TypeApplications #-}

module ColoniesLib  (
    Colony (..), 
    Runtime (..),
    Conditions (..),
    ProcessSpec (..),
    createColony,
    addColony,
    getColony,
    getColonies,
    createRuntime,
    addRuntime,
    approveRuntime,
    createConditions,
    createProcessSpec,
    createEmptyProcess,
    addEnv,
    submit,
    assign,
    getCmd,
    getArgs,
    addAttribute,
    createAttribute,
    close,
    failed
  ) where 

import Data.Aeson as JSON
import GHC.Generics
import Network.HTTP.Simple
import Data.ByteString.Base64 as Base64
import qualified Data.ByteString as B
import qualified Data.ByteString.Internal as BI
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Internal as BLI
import Data.Text as T
import Data.Text.IO as T
import Data.Text.Lazy as TL
import Data.Text.Lazy.Encoding as TL
import Data.Text.Lazy.IO as TL
import Data.Map as M
import Data.Typeable
import Control.Exception (try)
import Prelude hiding (error) 
import GHC.Records (getField)
import CryptoLib

-- Data types declaration
data RPCMsg = RPCMsg { signature   :: T.Text,
                       payloadtype :: T.Text,
                       payload     :: T.Text
                     } deriving (Show, Generic)
instance FromJSON RPCMsg
instance ToJSON RPCMsg

data RPCReplyMsg = RPCReplyMsg { error       :: Bool,
                                 payloadtype :: T.Text,
                                 payload     :: T.Text
                               } deriving (Show, Generic)
instance FromJSON RPCReplyMsg
instance ToJSON RPCReplyMsg

data Err = Err deriving (Show, Eq)

data RPCOperations = AddColonyRPCMsg { colony  :: Colony, msgtype :: T.Text } 
                   | GetColonyRPCMsg { colonyid :: T.Text, msgtype :: T.Text }
                   | GetColoniesRPCMsg { msgtype :: T.Text }
                   | AddRuntimeRPCMsg { runtime :: Runtime, msgtype :: T.Text }
                   | ApproveRuntimeRPCMsg { runtimeid :: T.Text, msgtype :: T.Text }
                   | SubmitProcessSpecRPCMsg { spec :: ProcessSpec, msgtype :: T.Text }
                   | AssignProcessRPCMsg { colonyid :: T.Text, msgtype :: T.Text }
                   | AddAttributeRPCMsg { attribute :: Attribute, msgtype :: T.Text }
                   | CloseSuccessfulRPCMsg { processid :: T.Text, msgtype :: T.Text }
                   | CloseFailedRPCMsg { processid :: T.Text, msgtype :: T.Text }
                       deriving (Show, Generic)
instance FromJSON RPCOperations 
instance ToJSON RPCOperations 

data Colony = Colony { colonyid :: T.Text,
                       name :: T.Text } deriving (Show, Generic, Eq)
instance FromJSON Colony
instance ToJSON Colony

data Runtime = Runtime { runtimeid :: T.Text,
                         runtimetype :: T.Text,
                         name :: T.Text,
                         colonyid :: T.Text,
                         cpu :: T.Text,
                         cores :: Int, 
                         mem :: Int,
                         gpu :: T.Text,
                         gpus :: Int,
                         state :: Int,
                         commissiontime :: T.Text,
                         lastheardfromtime :: T.Text } deriving (Show, Generic, Eq)
instance FromJSON Runtime 
instance ToJSON Runtime 

data Conditions = Conditions { colonyid :: T.Text, 
                               runtimeids :: [T.Text],
                               runtimetype :: T.Text,
                               mem :: Int, 
                               cores :: Int,
                               gpus :: Int,
                               dependencies :: [T.Text] } deriving (Show, Generic, Eq)
instance FromJSON Conditions 
instance ToJSON Conditions 

data ProcessSpec = ProcessSpec { name :: T.Text,
                                 image :: T.Text,
                                 cmd :: T.Text,
                                 args :: [T.Text],
                                 volumes :: [T.Text],
                                 ports :: [T.Text],
                                 priority :: Int, 
                                 maxexectime :: Int,
                                 maxretries :: Int,
                                 conditions :: Conditions,
                                 env :: Map T.Text T.Text } deriving (Show, Generic, Eq)
instance FromJSON ProcessSpec 
instance ToJSON ProcessSpec 

data Attribute = Attribute { attributeid :: T.Text,
                             targetid :: T.Text,
                             targetcolonyid :: T.Text,
                             targetprocessgraphid :: T.Text,
                             attributetype :: Int,
                             key :: T.Text,
                             value :: T.Text } deriving (Show, Generic, Eq)
instance FromJSON Attribute 
instance ToJSON Attribute 

data Process = Process { processid :: T.Text,
                         assignedruntimeid :: T.Text,
                         isassigned :: Bool,
                         state :: Int,
                         submissiontime :: T.Text,
                         starttime :: T.Text,
                         endtime :: T.Text, 
                         deadline :: T.Text,
                         retries :: Int,
                         attributes :: [Attribute],
                         spec :: ProcessSpec,
                         waitforparents :: Bool,
                         parents :: [T.Text],
                         children :: [T.Text],
                         processgraphid :: T.Text} deriving (Show, Generic, Eq)
instance FromJSON Process 
instance ToJSON Process

-- RPC functions  
checkError :: Either HttpException (Response BLI.ByteString) -> Maybe Err 
checkError e = case e of
    Left e -> Just Err
    Right resp -> Nothing 
    
sendRPCMsg :: RPCOperations -> String -> String -> IO (Either HttpException (Response BLI.ByteString))
sendRPCMsg rpcMsg host key = do
    let jsonEncoded = JSON.encode rpcMsg 
    let base64Encoded = Base64.encode $ BL.toStrict jsonEncoded
    signature <- sign (BI.unpackChars base64Encoded) key
    let payloadType = getField @"msgtype" rpcMsg
    let rpcMsg = RPCMsg { signature = T.pack (signature), 
                          payload = T.pack (BI.unpackChars (base64Encoded)), 
                          payloadtype = payloadType } 
    let jsonEncodedRPCMsg =  JSON.encode rpcMsg

    request' <- parseRequest host 
    let request
            = setRequestMethod "POST"
            $ setRequestPath "/api"
            $ setRequestBodyLBS jsonEncodedRPCMsg
            $ setRequestSecure False 
            $ request'
    try $ httpLBS request

-- Parser functions
parsePayload :: T.Text -> Either a BI.ByteString -> Maybe (T.Text, BLI.ByteString)
parsePayload payloadType jsonEncoded = do
    case jsonEncoded of
        Left e -> Nothing 
        Right j -> do let json = j
                      let lazyJSON = BL.fromStrict json
                      Just (payloadType, lazyJSON)

parseResponse :: Either HttpException (Response BLI.ByteString) -> Maybe (T.Text, BLI.ByteString)
parseResponse e = case e of
    Left e -> Nothing
    Right resp -> do let json = getResponseBody resp
                     let rpcReplyMsg = JSON.eitherDecode json :: Either String RPCReplyMsg
                     case rpcReplyMsg of
                          Left e -> Nothing
                          Right p -> do let payload = getField @"payload" p 
                                        let payloadType = getField @"payloadtype" p
                                        let jsonEncoded = Base64.decode $ BI.packChars $ T.unpack payload
                                        parsePayload payloadType jsonEncoded

parseRuntime :: Maybe (T.Text, BLI.ByteString) -> Maybe Runtime 
parseRuntime res = do  
    case res of
        Nothing -> Nothing
        Just tuple -> do let payloadType = fst tuple
                         let json = snd tuple 
                         JSON.decode json :: Maybe Runtime 

parseColony :: Maybe (T.Text, BLI.ByteString) -> Maybe Colony
parseColony res = do  
    case res of
        Nothing -> Nothing
        Just tuple -> do let payloadType = fst tuple
                         let json = snd tuple 
                         JSON.decode json :: Maybe Colony

parseColonies :: Maybe (T.Text, BLI.ByteString) -> Maybe [Colony]
parseColonies res = do  
    case res of
        Nothing -> Nothing
        Just tuple -> do let payloadType = fst tuple
                         let json = snd tuple 
                         JSON.decode json :: Maybe [Colony]

parseProcess :: Maybe (T.Text, BLI.ByteString) -> Maybe Process
parseProcess res = do  
    case res of
        Nothing -> Nothing
        Just tuple -> do let payloadType = fst tuple
                         let json = snd tuple 
                         JSON.decode json :: Maybe Process 

parseAttribute :: Maybe (T.Text, BLI.ByteString) -> Maybe Attribute 
parseAttribute res = do  
    case res of
        Nothing -> Nothing
        Just tuple -> do let payloadType = fst tuple
                         let json = snd tuple 
                         JSON.decode json :: Maybe Attribute 

-- API functions 
createColony :: String -> String -> Colony
createColony colonyid name = 
    Colony { colonyid = T.pack colonyid, name = T.pack name }

addColony :: Colony -> String -> String -> IO (Maybe Colony)
addColony colony host key = do
    resp <- sendRPCMsg AddColonyRPCMsg { colony = colony, msgtype = "addcolonymsg" } host key 
    return $ parseColony $ parseResponse resp 

getColony :: String -> String -> String -> IO (Maybe Colony) 
getColony colonyId host key = do
    resp <- sendRPCMsg GetColonyRPCMsg { colonyid = T.pack colonyId, msgtype = "getcolonymsg" } host key 
    return $ parseColony $ parseResponse resp 

getColonies :: String -> String -> IO (Maybe [Colony]) 
getColonies host key = do
    resp <- sendRPCMsg GetColoniesRPCMsg { msgtype = "getcoloniesmsg" } host key 
    return $ parseColonies $ parseResponse resp 

createRuntime :: String -> String -> String -> Runtime 
createRuntime runtimeType runtimeId colonyId = 
    Runtime { runtimeid = T.pack runtimeId, 
              runtimetype = T.pack runtimeType, 
              name ="test_name", 
              colonyid = T.pack colonyId,
              cpu = "",
              cores = 0,
              mem = 0,
              gpu = "",
              gpus = 0,
              state = 0,
              commissiontime = "2022-07-10T13:32:17.117545582+02:00",
              lastheardfromtime = "2022-07-10T13:32:17.117545582+02:00"}

addRuntime :: Runtime -> String -> String -> IO (Maybe Runtime)
addRuntime runtime host key = do
    resp <- sendRPCMsg AddRuntimeRPCMsg { runtime = runtime, msgtype = "addruntimemsg" } host key 
    return $ parseRuntime $ parseResponse resp 

approveRuntime :: String -> String -> String -> IO (Maybe Err)
approveRuntime runtimeId host key = do
    resp <- sendRPCMsg ApproveRuntimeRPCMsg { runtimeid = T.pack runtimeId, msgtype = "approveruntimemsg" } host key 
    return $ checkError resp

createConditions :: String -> String -> [String] -> Conditions 
createConditions colonyId runtimeType dependencies =  
    Conditions { colonyid = T.pack colonyId,
                            runtimeids = [],
                            runtimetype = T.pack runtimeType,
                            mem = 0, 
                            cores = 0,
                            gpus = 0,
                            dependencies = fmap (T.pack) dependencies }

createProcessSpec :: String -> String -> [String] -> Int -> Int -> Conditions -> ProcessSpec 
createProcessSpec name cmd args maxexectime maxretries conditions =
  ProcessSpec { name = T.pack name,
                image = "",
                cmd = T.pack cmd,
                args = fmap (T.pack) args,
                volumes = [],
                ports = [],
                priority = 0,
                maxexectime = maxexectime,
                maxretries = maxretries,
                conditions = conditions,
                env = M.empty } 

createProcessSpecWithEnv :: String -> String -> [String] -> Int -> Int -> Map T.Text T.Text -> Conditions -> ProcessSpec 
createProcessSpecWithEnv name cmd args maxexectime maxretries env conditions =
  ProcessSpec { name = T.pack name,
                image = "",
                cmd = T.pack cmd,
                args = fmap (T.pack) args,
                volumes = [],
                ports = [],
                priority = 0,
                maxexectime = maxexectime,
                maxretries = maxretries,
                conditions = conditions,
                env = env} 

addEnv :: ProcessSpec -> String -> String -> ProcessSpec
addEnv spec key value = do let name = T.unpack (getField @"name" spec)
                           let cmd = T.unpack (getField @"cmd" spec)
                           let args = fmap (T.unpack) (getField @"args" spec)
                           let maxexectime = getField @"maxexectime" spec
                           let maxretries = getField @"maxretries" spec
                           let conditions = getField @"conditions" spec
                           let env = getField @"env" spec
                           let textKey = T.pack key
                           let textValue = T.pack value
                           let newEnv = M.insert textKey textValue env
                           createProcessSpecWithEnv name cmd args maxexectime maxretries newEnv conditions  

createEmptyProcess :: Process
createEmptyProcess = 
    Process { processid = "",
              assignedruntimeid  = "",
              isassigned = False,
              state = -1,
              submissiontime = "",
              starttime = "",
              endtime = "", 
              deadline = "",
              retries = -1,
              attributes = [],
              spec = ProcessSpec { name = "",
                                   image = "",
                                   cmd = "",
                                   args = [],
                                   volumes = [],
                                   ports = [],
                                   priority = 0,
                                   maxexectime = -1,
                                   maxretries = -1,
                                   conditions = Conditions { colonyid = "",
                                                             runtimeids = [],
                                                             runtimetype = "",
                                                             mem = 0, 
                                                             cores = 0,
                                                             gpus = 0,
                                                             dependencies = [] },
                                   env = M.empty },
              waitforparents = False,
              parents = [],
              children = [],
              processgraphid  = ""}

submit :: ProcessSpec -> String -> String -> IO (Maybe Process)
submit spec host key = do 
    resp <- sendRPCMsg SubmitProcessSpecRPCMsg { spec = spec, msgtype = "submitprocessespecmsg" } host key 
    return $ parseProcess $ parseResponse resp 

assign :: String -> String -> String -> IO (Maybe Process)
assign colonyId host key = do 
    resp <- sendRPCMsg AssignProcessRPCMsg { colonyid = T.pack colonyId, msgtype = "assignprocessmsg" } host key 
    return $ parseProcess $ parseResponse resp 

getCmd :: Process -> IO String
getCmd process = do 
    let spec = getField @"spec" process 
    return $ T.unpack $ getField @"cmd" spec 

getArgs :: Process -> IO [String]
getArgs process = do 
    let spec = getField @"spec" process 
    return $ fmap (T.unpack) $ getField @"args" spec

createAttribute :: Process -> String -> String -> Attribute 
createAttribute process key value = 
    Attribute { attributeid = "",
                targetid = getField @"processid" process,
                targetcolonyid = getField @"colonyid" $ getField @"conditions" $ getField @"spec" process,
                targetprocessgraphid = "",
                attributetype = 1,
                key = T.pack key,
                value = T.pack value } 

addAttribute :: Attribute -> String -> String -> IO (Maybe Attribute)
addAttribute attr host key = do
    resp <- sendRPCMsg AddAttributeRPCMsg { attribute = attr, msgtype = "addattributemsg" } host key 
    return $ parseAttribute $ parseResponse resp 

close :: Process -> String -> String -> IO (Maybe Err)
close process host key = do
    resp <- sendRPCMsg CloseSuccessfulRPCMsg { processid = getField @"processid" process, msgtype = "closesuccessfulmsg" } host key 
    return $ checkError resp 

failed :: Process -> String -> String -> IO (Maybe Err)
failed process host key = do
    resp <- sendRPCMsg CloseFailedRPCMsg { processid = getField @"processid" process, msgtype = "closefailedmsg" } host key 
    return $ checkError resp 


app/Main.hs
module Main where

import CryptoLib
import ColoniesLib
import Control.Concurrent (threadDelay)
import Control.Monad (forever)

colonyId = "4787a5071856a4acf702b2ffcea422e3237a679c681314113d86139461290cf4"
runtimePrvKey = "ddf7f7791208083b6a9ed975a72684f6406a269cfa36f1b1c32045c0a71fff05"
host = "http://localhost:50080"

fib 0 = 0
fib 1 = 1
fib n = fib (n-1) + fib (n-2)

worker :: IO ()
worker = do
    -- Connect to the Colonies server and try to assign a process to execute from the job queue
    maybeProcess <- assign colonyId host runtimePrvKey
    if maybeProcess == Nothing then
        threadDelay 1000000 -- Job queue is empty, wait 1 second and try again ...
    else do
        -- Parse process parameters
        let process = maybe createEmptyProcess id maybeProcess
        cmd <- getCmd process
        args <- getArgs process
        if cmd == "fibonacci" then do
            print "Got a Fibonacci task"
            let n = read $ head args :: Integer
            let f = fib n 
            print $ "-> fib " ++ show n ++ "=" ++ show f
            -- Create a an attribute 
            let attr = createAttribute process "output" $ show f
            -- Connect to the Colnies server and add the add attribute to the process object
            -- Note: addAddtribute will return Nothing in case of error, else a Just Attribute
            maybeAddedAttr <- addAttribute attr host runtimePrvKey
            close process host runtimePrvKey 
            print "Done calculating Fibonacci"
        else
            print "Invalid cmd"

main :: IO ()
main = forever worker 
